### PR TITLE
feat(pt_expt): DPA4 descriptor and fitting (training + export)

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
PR-2 of the DPA4/SeZM porting series (PR-1: #5515, dpmodel core). Makes DPA4 trainable and torch.export-compatible on the pt_expt backend.

## What's included

- **Descriptor wrapper** (`deepmd/pt_expt/descriptor/dpa4.py`): thin `@torch_module` `DescrptDPA4` registered as `SeZM`/`sezm`/`DPA4`/`dpa4`, plus explicit converter registrations for the two dpmodel classes that cannot be auto-wrapped via serialize round-trip: `WignerDCalculator` (deserialize raises by design — rebuilt from `lmax`/`eps`/`precision`) and `SwiGLU` (parameter-free, no serialize).
- **Trainable-weight promotion**: dpmodel stores DPA4 weights as bare numpy arrays (mirroring pt SeZM parameter names), which the generic ndarray rule registers as non-trainable buffers. A localized `_TRAINABLE_ATTRS` table re-registers exactly the attributes that are `nn.Parameter` in the pt SeZM implementation, after `__init__`/`deserialize`. Two supporting infra fixes: `_try_convert_list` now handles Optional-module lists (`SO2Convolution.non_linearities` ends with `None`), and `xp_asarray_nodetach` (`deepmd/dpmodel/array_api.py`) replaces `xp.asarray` on weight attributes in dpa4 dpmodel code so torch autograd is not silently detached.
- **Fitting wrapper** (`deepmd/pt_expt/fitting/dpa4_ener.py`): `SeZMEnergyFittingNet` (`dpa4_ener`/`sezm_ener`), `GLUFittingNet`, and a `SeZMNetworkCollection` wrapper mirroring the `NetworkCollection` ModuleDict pattern so GLU weights are optimizer-visible.
- **Model assembly** (`deepmd/pt_expt/model/get_model.py`): `get_sezm_model` dispatched for model type `dpa4`/`sezm`/`SeZM`/`DPA4`, mirroring pt's config handling (descriptor/fitting type defaults, `pair_exclude_types` reconciliation). Fail-fast `NotImplementedError` guards for pt-only extensions: spin, `bridging_method != none`, `lora`, `use_compile`, `preset_out_bias`. `enable_tf32` is accepted and ignored (pt_expt always runs "highest" matmul precision; documented).

## Tests

- pt_expt trio (`source/tests/pt_expt/descriptor/test_dpa4.py`): consistency (serde round-trip + dpmodel reference), `torch.export.export`, `make_fx` through forward + `autograd.grad`; fitting analogs in `source/tests/pt_expt/fitting/test_dpa4_ener.py`.
- **Parameter-gradient parity vs pt** (`source/tests/pt/model/test_dpa4_ptexpt_grad_parity.py`): weight-copied pt SeZM vs pt_expt, quadratic loss, every parameter's gradient compared 1:1 via serialize-tree alignment (83/83 parameters, no dead parameters, max deviation ~3e-11 rel from fp64 summation order).
- Model assembly tests (`source/tests/pt_expt/model/test_get_model_dpa4.py`): aliases, defaults, exclude-types branches, all NIE guards.
- Training end-to-end (`source/tests/pt_expt/test_training.py::test_training_loop_dpa4`) on the water fixture through the full trainer.
- Cross-backend consistency rows enabled for pt_expt (`source/tests/consistent/`): 20 new comparisons pass at existing harness tolerances.
- PR-1 parity suites stay green (381 passed) including for the small dpmodel fixes this PR needed (serialize via `to_numpy_array`, explicit-dtype zeros, ParameterList-safe `_load_variables`, `charge_spin` kwarg for atomic-model interface compatibility).

## Known limitations

- No freeze/`.pt2` export, no Python `DeepEval`, no C++/LAMMPS inference — that is PR-3 of the series.
- No multitask: `share_params` raises `NotImplementedError`.
- No model-dict deserialization of pt `SeZMModel` checkpoints (PR-3).
- pt-only SeZM extensions fail fast rather than being ported: spin, bridging/ZBL, LoRA, `use_compile`, `preset_out_bias`; `enable_tf32` ignored (numerically conservative).
- `_try_convert_list`'s ParameterList branch does not respect `trainable=False` for list-valued weights (pre-existing backend-wide behavior).
- CUDA validated at tolerance level only in CI; performance vs the pt Triton implementation is unbenchmarked.

## Summary by CodeRabbit

* **New Features**
  * Added PyTorch-experimental DPA4/SeZM descriptor and energy-fitting components, with public package exports.
  * Added EnergyModel factory/dispatch to build DPA4/SeZM models from configs.

* **Improvements**
  * Enhanced DPA4/SeZM runtime handling to keep gradient connectivity and ensure device-consistent tensor placement.
  * Updated DPA4 descriptor interface to accept optional `charge_spin` for compatibility.

* **Tests**
  * Expanded experimental coverage with serialization/parity checks, FX tracing/export validation, trainable-parameter behavior tests, and descriptor/fitting gradient parity verification.

### CODE REVIEW

File: source/tests/consistent/fitting/test_dpa4_ener.py [hunk 1]
```diff
@@ -30,6 +31,15 @@
     from deepmd.pt.utils.env import DEVICE as PT_DEVICE
 else:
     SeZMEnerFittingPT = None
+if INSTALLED_PT_EXPT:
+    import torch
+
+    from deepmd.pt_expt.fitting.dpa4_ener import (
+        SeZMEnergyFittingNet as SeZMEnerFittingPTExpt,
+    )
+    from deepmd.pt_expt.utils.env import DEVICE as PT_EXPT_DEVICE
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/consistent/fitting/test_dpa4_ener.py:40]: The value-import of `DEVICE` is the established pattern of the consistency-test harness: `source/tests/consistent/fitting/test_ener.py` (lines 35 and 42) does exactly `from deepmd.pt.utils.env import DEVICE as PT_DEVICE` / `from deepmd.pt_expt.utils.env import DEVICE as PT_EXPT_DEVICE`, as do the other consistent tests. `DEVICE` is computed once at import time in `deepmd/pt/utils/env.py` (from `torch.cuda.is_available()` / `LOCAL_RANK`) and is never reassigned at runtime, so the value-import cannot go stale. Switching this one file to module-attribute access would diverge from the harness convention without changing behavior.

File: source/tests/pt_expt/descriptor/test_dpa4.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        nlist = torch.tensor(self.nlist, dtype=int, device=self.device)
+        mapping = (
+            torch.tensor(self.mapping, dtype=int, device=self.device)
+            if use_mapping
+            else None
+        )
+        rd0 = dd0(coord_ext, atype_ext, nlist, mapping)[0]
+        # serialization round-trip within pt_expt
+        dd1 = DescrptDPA4.deserialize(dd0.serialize())
+        rd1 = dd1(coord_ext, atype_ext, nlist, mapping)[0]
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/descriptor/test_dpa4.py:73]: `dd1` is already on `self.device`: `DescrptDPA4.deserialize` goes through `dpmodel_setattr` in `deepmd/pt_expt/common.py`, which materializes every tensor with `torch.as_tensor(value, device=env.DEVICE)` (line 299; same for buffers at line 208). This test sets `self.device = env.DEVICE` in `setup_method`, so the deserialized module and the inputs are on the same device by construction — including on CUDA. Adding `.to(self.device)` would be a no-op.

File: source/tests/pt_expt/test_training.py [hunk 3]
```diff
@@ -284,6 +311,15 @@ def test_training_loop(self) -> None:
...
         self._run_training(config)
 
+    def test_training_loop_dpa4(self) -> None:
+        """Run a few DPA4/SeZM training steps (model type "dpa4" dispatch)."""
+        config = _make_config(self.data_dir, numb_steps=5)
+        config["model"] = copy.deepcopy(_MODEL_DPA4)
+        config = update_deepmd_input(config, warning=False)
+        config = normalize(config)
+        self.assertEqual(config["model"]["type"], "dpa4")
+        self._run_training(config)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/test_training.py:321]: None of the existing training tests in `source/tests/pt_expt/test_training.py` (`test_training_loop`, `test_training_loop_compiled`, `test_training_loop_compiled_silu`, ...) use a per-test timeout; the suite relies on the global `timeout = 1800` configured in `pyproject.toml` (pytest-timeout). `test_training_loop_dpa4` runs the same 5-step smoke config as its neighbors and completes in well under a minute. Adding a 60s cap to only this one test would be inconsistent with the rest of the file; introducing per-test timeouts file-wide is out of scope for this PR.

File: source/tests/pt/model/test_dpa4_ptexpt_grad_parity.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        kwargs.update(overrides)
+        pt_mod = DescrptSeZM(**kwargs).double()
+        # several projections are zero-initialized; perturb for nonzero
+        # output and weight-dependent gradients everywhere
+        rng = np.random.default_rng(2150)
+        with torch.no_grad():
+            for p in pt_mod.parameters():
+                p += to_pt(0.05 * rng.normal(size=tuple(p.shape)))
+        expt_mod = DescrptDPA4.deserialize(pt_mod.serialize())
+        return pt_mod, expt_mod
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt/model/test_dpa4_ptexpt_grad_parity.py:137]: Both models are constructed on `PT_DEVICE` already. The pt `DescrptSeZM` sub-modules create all their parameters on `env.DEVICE` at construction (e.g. `sezm_nn/so3.py`: `self.device = env.DEVICE` then `nn.Parameter(torch.empty(..., device=self.device, ...))`; same in `so2.py`, `norm.py`, etc.), and `.double()` only changes dtype, not device. `expt_mod = DescrptDPA4.deserialize(...)` goes through `dpmodel_setattr` in `deepmd/pt_expt/common.py`, which places every tensor with `torch.as_tensor(value, device=env.DEVICE)` (line 299). `PT_DEVICE = pt_env.DEVICE` and the pt/pt_expt env modules compute `DEVICE` identically, so models and `to_pt(...)` inputs land on the same device — this is exactly why the file carries the non-CPU tolerance gates.

File: deepmd/pt_expt/descriptor/dpa4.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    "SeZMTypeEmbedding": ("adam_type_embedding",),
+    # descriptor-level FiLM strengths
+    "DescrptDPA4": ("film_scale_strength_log", "film_shift_strength_log"),
+}
+
+
+def _promote_trainable(module: torch.nn.Module, names: tuple[str, ...]) -> None:
+    """Re-register the given float buffers of *module* as Parameters."""
+    if not getattr(module, "trainable", True):
+        return
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/descriptor/dpa4.py:97]: addressed in 374930088

**Review by OutisLi — APPROVED** [review]
